### ISSUE TITLE
Dynamic imports started failing in 0.26.0

### ISSUE THREAD

**Issue by purepear** [vitest-dev/vitest#2552]
### Describe the bug

I'm having problems with dynamic imports
* everything was working in 0.25.8
* everything is working in 0.26.2 when I run **individual** test files (`npx vitest run individualFile.test.js`)

The problem happens when I test multiple test files that have modules which dynamically import other modules:
```js
// For example
await import(`../../${sth}/${sthElse}.js`)
```

The problem doesn't happen all the time - sometimes the tests pass, sometimes they fails and it's not always the exact same tests that fail. It feels like there's some sort of race condition between the workers when they start loading dynamic modules.

I made sure to await in all places 
```js
test('sth', async () => {
  // render a component that has `async created()` that load dynamic modules 

  await flushPromises() 
  await vi.dynamicImportSettled()
  
  // expect the side effects of the dynamic imports to have happened
})  
```

### Reproduction

It's difficult to put together a reproduction because of the complexity of the project.
Any tips that would help narrow this down are welcome. 🙏 

I tried creating a repro but it's still not ready - couldn't reproduce the flaky failures.

However I bumped into **an interesting problem which may or may not be related**:
* starting the tests in watch mode works `npx vitest`
* running the tests once hangs (CPU 100%) `npx vitest run`

https://stackblitz.com/edit/vitest-dynamic-imports-issues

### System Info

```shell
System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 1.34 GB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.19.0 - ~/.volta/tools/image/node/16.19.0/bin/node
    Yarn: 1.22.18 - ~/.volta/tools/image/yarn/1.22.18/bin/yarn
    npm: 8.19.3 - ~/.volta/tools/image/node/16.19.0/bin/npm
  Browsers:
    Chrome: 108.0.5359.124
    Firefox: 107.0.1
    Safari: 16.2
  npmPackages:
    @vitejs/plugin-vue: ^3.2.0 => 3.2.0
    vite: ^3.2.5 => 3.2.5
    vitest: ^0.26.2 => 0.26.2
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by cexbrayat** [comment]
We're also facing a similar issue, as @purepear, especially on CI. I tried to reproduce the issue in isolation, but it seems to be random (which confirms what happens on our CI jobs: we re-run them and they succeed).

The only tests that randomly fail are those using `vi.dynamicImportSettled()`, and they started failing with v0.26.0 (and are still failing with v0.26.2).

**Comment by ahnpnl** [comment]
I also encountered the same issue. Rolling back to **0.25.8** fixes the problem.

**Comment by sheremet-va** [comment]
> running the tests once hangs (CPU 100%) npx vitest run

I noticed this in stackblitz, but cannot reproduce locally. Do you also this this locally?

**Comment by purepear** [comment]
Hey @sheremet-va, I tested your [not yet merged PR](https://github.com/vitest-dev/vitest/pull/2560) and it seems to work fine 🎉 
Thank you so much!

For others having similar problems: after updating I still had to add `await vi.dynamicImportSettled()` in few places that were previously working without it in `0.25.8`(not sure how things worked before - it looks like I for got to await in the first place)... but  your PR seems to fix the flakiness in `0.26.2`. Looking forward to the next release :) 


Re stackblitz:
> I noticed this in stackblitz, but cannot reproduce locally. Do you also this this locally?

I just transferred the stackblitz code locally and it runs fine. Seem like an unrelated stackblitz-specific issue.
